internal/commands/tunnel/tunnel.go

```diff
@@ -36,44 +80,161 @@ func NewCmdTunnel() *cobra.Command {
+	o.Region = viper.GetString("aws-region")
+
+	if o.Profile == "" {
+		o.Profile = viper.GetString("aws_profile")
+	}
+
+	if o.Region == "" {
+		o.Region = viper.GetString("aws_region")
+	}
+
+	if o.PrivateKeyFile == "" {
+		home, _ := os.UserHomeDir()
+		o.PrivateKeyFile = fmt.Sprintf("%s/.ssh/id_rsa", home)
 	}
+
+	if o.PublicKeyFile == "" {
+		home, _ := os.UserHomeDir()
+		o.PublicKeyFile = fmt.Sprintf("%s/.ssh/id_rsa.pub", home)
+	}
+
+	return nil
 }
 
-func tunnelDown(dCtx *daemon.Context) error {
-	p, err := dCtx.Search()
-	if err != nil {
-		return fmt.Errorf("search for daemon process: %w", err)
+func (o *TunnelOptions) Validate() error {
+	if len(o.Env) == 0 {
+		return fmt.Errorf("env must be specified")
 	}
-	pterm.Info.Printf("killing daemon process(pid: %d)\n", p.Pid)
 
-	if err := p.Signal(syscall.SIGTERM); err != nil {
-		return fmt.Errorf("kill daemon process(pid: %d): %w", p.Pid, err)
+	if len(o.Profile) == 0 {
+		return fmt.Errorf("AWS profile must be specified")
 	}
-	return os.Remove(dCtx.PidFileName)
+
+	if len(o.Region) == 0 {
+		return fmt.Errorf("AWS region must be specified")
+	}
+	return nil
 }
 
-func daemonRunning(dCtx *daemon.Context) (process *os.Process, running bool, err error) {
-	p, err := dCtx.Search()
+func (o *TunnelOptions) Run(cmd *cobra.Command) error {
+	pterm.DefaultSection.Printfln("Sending SSH public key")
+
+	sess, err := utils.GetSession(&utils.SessionConfig{
+		Region:  o.Region,
+		Profile: o.Profile,
+	})
 	if err != nil {
-		if os.IsNotExist(err) {
-			return nil, false, nil
-		}
-		return nil, false, fmt.Errorf("search daemon process: %w", err)
+		return fmt.Errorf("can't get AWS session: %w", err)
+	}
+
+	logrus.Debug("getting AWS session")
+
+	to, err := getTerraformOutput(sess, o.Env)
+	if err != nil {
+		return fmt.Errorf("can't get forward config: %w", err)
+	}
+
+	logrus.Debug("getting bastion instance ID")
+
+	logrus.Debugf("public key path: %s", o.PublicKeyFile)
+
+	err = sendSSHPublicKey(to.BastionInstanceID.Value, getPublicKey(o.PublicKeyFile), sess)
+	if err != nil {
+		logrus.Error("sending user SSH public key")
+		return err
 	}
-	err = p.Signal(syscall.Signal(0))
+
+	logrus.Debug("sending user SSH public key")
+
+	sigCh := make(chan os.Signal, 1)
+	signal.Notify(sigCh, os.Interrupt, syscall.SIGTERM, syscall.SIGHUP)
+
+	ctx, cancel := context.WithCancel(cmd.Context())
+	defer cancel()
+
+	pterm.DefaultSection.Printfln("Running SSH Tunnel Up")
+
+	hosts := getForwardConfig(to)
+
+	logrus.Debug("getting SSH forward config")
+
+	logrus.Debugf("hosts: %s", hosts)
+
+	if len(hosts) == 0 {
+		return fmt.Errorf("can't tunnel up: forward config is not valid")
+	}
+
+	logrus.Debug("forwarding config is valid")
+
+	_ = killDaemon(daemonContext(ctx))
+
+	p, err := daemonContext(ctx).Reborn()
 	if err != nil {
-		return p, false, nil
+		return fmt.Errorf("restarted tunnel process: %w", err)
+	}
+	if p != nil {
+		pterm.Info.Printf("tunnel is up(pid: %d)\n", p.Pid)
```

Comment:
Can we hide the "(pid: %d)" and show it only if debug is on? If not, let's add a space after up at least. 

Also, if possible for "tunnel" command let's also run "tunnel status" so we see the ports.

Reply:
@AutomationD,
I can display ports only before running forwarding because when running the forwarding app go in the daemon.

So will it be okay?

Reply:
Sounds good 👍🏻